src/hotspot/share/opto/loopConditionalPropagation.cpp

```diff
@@ -0,0 +1,1213 @@
+
+  For a Phi phi at a region r:
+
+  type'[phi, r] = meet(type'[phi->in(1), r->in(1)], type'[phi->in(2), r->in(2)]...)
+
+  Once a type is narrowed, uses are enqueued and their types are
+  computed by calling the Value() methods. Value() methods retrieve
+  types from the type table, not the type' table. To address that
+  issue while leaving Value() methods unchanged, before calling
+  Value() at c, the type table is updated so:
+
+  type[n] = type'[n, c]
+
+  An exception is for Phi::Value which needs to retrieve the type of
+  nodes are various controls: there a new type(Node* n, Node* c)
+  method is used.
+
+  For most n and c, type'[n, c] is likely the same as type[n], the
+  type recorded in the global igvn table (that there shouldn't be many
+  nodes at only a few control for which we can narrow the type
+  down). As a consequence, the types'[n, c] table is implemented as:
+
+  - At c, narrowed down types are stored in a GrowableArray. Each
+    entry records the previous type at idom(c) and the narrowed down
+    type at c.
+
+  - The GrowableArray of type updates is recorded in a hash table
+    indexed by c. If there's no update at c, there's no entry in the
+    hash table.
+
+  This pass operates in 2 steps:
+
+  - it first iterates over the graph looking for conditions that
+    narrow the types of some nodes and propagate type updates to uses
+    until a fix point.
+
+  - it transforms the graph so newly found constant nodes are folded.
+
+*/
+
+#include "memory/resourceArea.hpp"
+#include "node.hpp"
+#include "precompiled.hpp"
+#include "opto/loopConditionalPropagation.hpp"
+#include "opto/callnode.hpp"
+#include "opto/movenode.hpp"
+#include "opto/opaquenode.hpp"
+
+
+bool PhaseConditionalPropagation::related_use(Node* u, Node* c) {
+  if (!_phase->has_node(u)) {
+    return false;
+  }
+  if (u->is_Phi()) {
+    if (u->in(0) ==  c) {
+      return true;
+    }
+    int iterations = _iterations;
+    if (_visited.test(u->in(0)->_idx)) {
+      _progress = true;
+      iterations = _iterations + 1;
+    }
+    // mark as needing processing either in the pass over the CFG (control not yet processed) or on the next one
+    _control_dependent_node[iterations%2].set(u->in(0)->_idx);
+    _control_dependent_node[iterations%2].set(u->_idx);
+    return false;
+  }
+  Node* u_c = _phase->ctrl_or_self(u);
+  if (!_phase->is_dominator(c, u_c) && (u->is_CFG() || !_phase->is_dominator(u_c, c))) {
+    return false;
+  }
+  if (!u->is_CFG() && u->in(0) != nullptr && u->in(0)->is_CFG() && !_phase->is_dominator(u->in(0), c)) {
+    // mark as needing processing either in the pass over the CFG (control not yet processed) or on the next one
+    assert(!_visited.test(u->in(0)->_idx), "");
+    _control_dependent_node[_iterations%2].set(u->in(0)->_idx);
+    _control_dependent_node[_iterations%2].set(u->_idx);
+    return false;
+  }
+  return true;
+}
+
+void PhaseConditionalPropagation::enqueue_uses(const Node* n, Node* c) {
```

Comment:
You're right. Having shared code to push uses that's shared across CCP, IGVN and this would be nice. There are 2 kind of issues, right: 1) missed opportunities that are otherwise harmless 2) inconsistencies that cause something to break. I would be happy if this avoided 2) initially. There's verification code that checks that no further progress can be made once we reached a fixed point.